### PR TITLE
Add wrapper for pilon

### PR DESCRIPTION
**Pull request by pvanheus**
pilon is a variant identification and assembly improvement tool from the Broad Institute, targeted at bacteria genomes (https://github.com/broadinstitute/pilon/wiki). It already exists in bioconda. This PR adds a tool XML file with associated test-data for pilon.

### CODE REVIEW

**Comment by bgruening** [comment]
@pvanheus is this for inclusion or for testing in regard to your mail to galaxy-dev?
Should we review it?

**Comment by pvanheus** [comment]
This is for inclusion - while we have a copy of this in the SANBI repo, better to have it in tools-iuc long term in my view.

File: tools/pilon/.shed.yml [hunk 1]
```diff
@@ -0,0 +1,12 @@
+categories:
+- Variant Analysis
+description: pilon is a tool for assembly improvement and variant analysis in bacteria
+long_description: |
+   Pilon is a software tool which can be used to:
+       * Automatically improve draft assemblies
+       * Find variation among strains, including large event detection
+   https://github.com/broadinstitute/pilon/wiki
+name: pilon
+owner: sanbi-uwc
```
Comment (bgruening) [inline tools/pilon/.shed.yml]: If this is a new tool, we would put this under the `iuc` account if you don't mind.

File: tools/pilon/.shed.yml [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+remote_repository_url: https://github.com/SANBI-SA/tools-sanbi-uwc/tree/master/tools/pilon
```
Comment (bgruening) [inline tools/pilon/.shed.yml]: This URL needs to be adopted.

File: tools/pilon/pilon.cwl [hunk 3]
```diff
@@ -0,0 +1,330 @@
+#!/usr/bin/env cwl-runner
```
Comment (bgruening) [inline tools/pilon/pilon.cwl]: Not needed currently.

File: tools/pilon/pilon.xml [hunk 4]
```diff
@@ -0,0 +1,284 @@
+<tool id="pilon" name="pilon" version="0.1">
+    <description>An automated genome assembly improvement and variant detection tool</description>
+    <requirements>
+      <requirement type="package" version="1.18">pilon</requirement>
+      <!--<requirement type="set_environment">CLASSPATH</requirement>-->
+      <!-- if CLASSPATH is set to include pilon jar, you can call pilon with
```
Comment (bgruening) [inline tools/pilon/pilon.xml]: Is this comment still needed?

File: tools/pilon/pilon.xml [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+           java com.simontuffs.onejar.Boot -->>
+    </requirements>
+    <stdio>
+        <exit_code range="1:" />
+    </stdio>
+    <command><![CDATA[
+      #for $input_bam in $input_bams
+        ln -f -s "${input_bam.metadata.bam_index}" "${input_bam}.bai" &&
```
Comment (bgruening) [inline tools/pilon/pilon.xml]: Here you creating bai files next to the original BAM file, isn't it. This pollutes the Galaxy database directory. Can you please link everything into the cwd.

File: tools/pilon/pilon.xml [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+      #end for
+      #if $options.selection_mode == "advanced"
+        #for $bam in $options.input_frag_bams + $options.input_jump_bams + $options.input_unpaired_bams
+          ln -s -f "${bam.metadata.bam_index}" "${bam}.bai" &&
```
Comment (bgruening) [inline tools/pilon/pilon.xml]: Same as above.

File: tools/pilon/pilon.xml [hunk 7]
```diff
@@ -0,0 +1,284 @@
...
+            <option value="bases">Individual bases and small indels</option>
+            <option value="gaps">Fill gaps</option>
+            <option value="local">Detect and fix local misassemblies</option>
+            <option value="none">Do none of these fixes (no FASTA will be written)</option>
+            <option value="amb">Fix ambigious bases in FASTA output (experimental)</option>
+            <option value="breaks">Allow local reassembly to open new gaps (experimental, requires local assembly fixing to be selected)</option>
+            <option value="novel">Assemble novel sequence from unaligned non-jump reads (experimental)</option>
+          </param>
+          <param name="diploid" label="Organism is diploid" type="boolean" checked="false"
+            help="Sample is from diploid organism; will eventually affect calling of heterozygous SNPs"
```
Comment (bgruening) [inline tools/pilon/pilon.xml]: We actually recommend to include the original parameter in the `help` section. You could also use the `argument=""` argument. Like here: https://github.com/galaxyproject/tools-iuc/blob/3c4a2b13b0f3a280de4f98f4f5e0dc29e10fc7a0/tools/ngsutils/bam_filter.xml#L74

`argument` will automatically add the original argument to the help and you do not need specify `name`. The cheetah variable will be what you have specified in `argument` just without trailing dashes.

File: tools/pilon/pilon.xml [hunk 8]
```diff
@@ -0,0 +1,284 @@
...
+      <data format="txt" label="Changes in FASTA from ${tool.name} on ${on_string}" name="output_changes">
+        <filter>changes</filter>
+      </data>
+      <data format="fasta" label="FASTA from ${tool.name} on ${on_string}" name="output_fasta">
+        <filter>options['selection_mode'] == 'default' or (options['selection_mode'] == 'advanced' and 'none' not in options['fixes'])</filter>
+      </data>
+      <data format="bed" label="Features from ${tool.name} on ${on_string} (BED format)" name="output_pilon_bed">
+        <filter>options['selection_mode'] == 'advanced' and options['tracks']</filter>
+      </data>
+      <data format="wig" label="${tool.name} changes track on ${on_string} (WIG format)" name="output_changes_wig">
```
Comment (bgruening) [inline tools/pilon/pilon.xml]: should the wig files all be in one collection?

File: tools/pilon/pilon.xml [hunk 9]
```diff
@@ -0,0 +1,284 @@
...
+  Note on **stray read filtering**
+
+  By default a pass is made through the input BAM files to identify stray pairs, that is,
+  those pairs in which both reads are aligned but not marked valid because they have
+  inconsistent orientation or separation. Identifying stray pairs can help fill gaps
+  and assemble larger insertions, especially of repeat content.  However, doing so
+  sometimes consumes considerable memory.
+    ]]></help>
+    <citations>
+        <citation type="bibtex">@article{Walker_2014,
```
Comment (bgruening) [inline tools/pilon/pilon.xml]: Is there a reason why you used bibtex and not the doi?

File: tools/pilon/pilon.xml [hunk 10]
```diff
@@ -0,0 +1,284 @@
...
+      <data format="wig" label="${tool.name} weighted MQ track on ${on_string}" name="output_weightedmq_wig">
+        <filter>options['selection_mode'] == 'advanced' and options['tracks']</filter>
+      </data>
+      <data format="wig" label="${tool.name} GC track on ${on_string}" name="output_gc_wig">
+        <filter>options['selection_mode'] == 'advanced' and options['tracks']</filter>
+      </data>
+    </outputs>
+    <tests>
+      <test>
+        <param name="reference_genome_source" value="history" />
```
Comment (bgruening) [inline tools/pilon/pilon.xml]: One test with advanced options activated would be great!

File: tools/pilon/.shed.yml (hunk 1 above)
Comment (pvanheus) [inline tools/pilon/.shed.yml]: Done.

File: tools/pilon/.shed.yml (hunk 2 above)
Comment (pvanheus) [inline tools/pilon/.shed.yml]: Pointed to the tools-iuc repo

File: tools/pilon/pilon.cwl (hunk 3 above)
Comment (pvanheus) [inline tools/pilon/pilon.cwl]: Removed.

File: tools/pilon/pilon.xml (hunk 4 above)
Comment (pvanheus) [inline tools/pilon/pilon.xml]: Removed.

File: tools/pilon/pilon.xml (hunk 5 above)
Comment (pvanheus) [inline tools/pilon/pilon.xml]: Changed location of link to working directory.

File: tools/pilon/pilon.xml (hunk 6 above)
Comment (pvanheus) [inline tools/pilon/pilon.xml]: Fixed as above.

File: tools/pilon/pilon.xml [hunk 11]
```diff
@@ -0,0 +1,284 @@
...
+          --K $options.kmersize
+          --mindepth $options.mindepth
+          --mingap $options.mingap
+          --minmq $options.minmq
+          --minqual $options.minqual
+          $options.nostrays
+        #end if
+        --threads \${GALAXY_SLOTS:-1}
+        --output pilon &&
+        if [ -f pilon.fasta ] ; then mv pilon.fasta $output_fasta ; fi &&
```
Comment (pvanheus) [inline tools/pilon/pilon.xml]: Now using from_work_dir instead of mv.

File: tools/pilon/pilon.xml (hunk 7 above)
Comment (pvanheus) [inline tools/pilon/pilon.xml]: Switched to argument="" and made sure help was up to date.

File: tools/pilon/pilon.xml (hunk 8 above)
Comment (pvanheus) [inline tools/pilon/pilon.xml]: Created list collection instead of individual files.

File: tools/pilon/pilon.xml (hunk 9 above)
Comment (pvanheus) [inline tools/pilon/pilon.xml]: At one point was getting timeouts with doi, these seem to be resolved now.

File: tools/pilon/pilon.xml (hunk 10 above)
Comment (pvanheus) [inline tools/pilon/pilon.xml]: Added a second test that uses advanced options (and revealed some bugs :) ).

Comment (bgruening) [inline tools/pilon/pilon.xml]: :)

File: tools/pilon/pilon.xml [hunk 12]
```diff
@@ -0,0 +1,320 @@
+<tool id="pilon" name="pilon" version="0.1">
+    <description>An automated genome assembly improvement and variant detection tool</description>
+    <requirements>
+      <requirement type="package" version="1.18">pilon</requirement>
+    </requirements>
+    <stdio>
+        <exit_code range="1:" />
```
Comment (bgruening) [inline tools/pilon/pilon.xml]: You can enhance this as well with `<command detect_errors="exit_code">` and remove your stdio tag.

File: tools/pilon/pilon.xml [hunk 13]
```diff
@@ -0,0 +1,320 @@
...
+        #if $options.unpaired_selection.unpaired_enabled == "yes"
+          #for $bamfile in $options.unpaired_selection.unpaired
+            ln -f -s "$bamfile" \$(basename $bamfile) &&
+            ln -f -s "$bamfile.metadata.bam_index" \$(basename $bamfile).bai &&
+          #end for
+        #end if
+      #end if
+        ln -s -f
+          #if $reference_genome.reference_genome_source == "history"
+            $reference_genome.history_item
```
Comment (bgruening) [inline tools/pilon/pilon.xml]: This should be quoted. One below as well.

**Comment by bgruening** [comment]
@pvanheus this looks great. I'm a little bit worried about the huge size of test data. Can we use `contains` or `md5` as test method to make them smaller?

**Comment by bgruening** [comment]
You are fast Peter! `+103 −140,339` :heart_eyes:

File: tools/pilon/pilon.xml (hunk 13 above)
Comment (bgruening) [inline tools/pilon/pilon.xml]: @pvanheus this two please and than I will hit the merge button.

**Comment by bgruening** [comment]
@pvanheus I can try to merge+squash or if you have time to squash your commits manually this would be great. It's to get rid of the test data completely.

**Comment by pvanheus** [comment]
Latest version is a squashed commit.

**Comment by bgruening** [comment]
Thanks a lot Peter!

**Comment by bgruening** [comment]
It's now alive. Thanks again!

**Pull request merged**